src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -517,65 +651,120 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
-            thisbm = thisbm & ~thislsb
-            thatbm = thatbm & ~thatlsb
-            thati += 1
-            thisi += 1
-          } else {
-            // condition below is due to 2 things:
-            // 1) no unsigned int compare on JVM
-            // 2) 0 (no lsb) should always be greater in comparison
-            if (unsignedCompare(thislsb - 1, thatlsb - 1)) {
-              val m = thiselems(thisi)
-              totalelems += m.size
-              merged(i) = m
+      case that: HashTrieMap[A, B1] =>
+        if ((this eq that) && merger.retainIdentical) this
+        else {
+          val thiselems = this.elems
+          val thatelems = that.elems
+          var thisbm = this.bitmap
+          var thatbm = that.bitmap
+
+          // determine the necessary size for the array
+          val subcount = Integer.bitCount(thisbm | thatbm)
+          var canBeThis = thisbm == (thisbm | thatbm)
+          var canBeThat = thatbm == (thisbm | thatbm)
+
+          // construct a new array of appropriate size
+          var merged: Array[HashMap[A, B1]] = if (canBeThis || canBeThat) null else new Array[HashMap[A, B1]](subcount)
+
+          // run through both bitmaps and add elements to it
+          var i = 0
+          var thisi = 0
+          var thati = 0
+          var totalelems = 0
+
+          // the result of merging in each iteration
+          // it mist be assigned in each iteration
+          // and pulled out here to avoid nulling on each cycle
+          // ideally it would be
+          // val mergeResult: HashMap[A, B1] = _
+          // inside the loop, but scala doesn't support the definite assignment rule that Java does
+          var mergeResult: HashMap[A, B1] = null
+
+          while (i < subcount) {
+            val thislsb = thisbm ^ (thisbm & (thisbm - 1))
+            val thatlsb = thatbm ^ (thatbm & (thatbm - 1))
+
+            // collision
+            if (thislsb == thatlsb) {
+              val thisAtI = thiselems(thisi)
+              val thatAtI = thatelems(thati)
+              //optimise for the special case of structurally shared maps, and the default merger
+              //or any merger that doesnt need to process identical sub-trees
+              if ((thisAtI eq thatAtI) && merger.retainIdentical) mergeResult = thisAtI
+              else mergeResult = thisAtI.merge0(thatAtI, level + 5, merger)
+              totalelems += mergeResult.size
               thisbm = thisbm & ~thislsb
-              thisi += 1
-            }
-            else {
-              val m = thatelems(thati)
-              totalelems += m.size
-              merged(i) = m
               thatbm = thatbm & ~thatlsb
               thati += 1
+              thisi += 1
+            } else {
+              // condition below is due to 2 things:
+              // 1) no unsigned int compare on JVM
+              // 2) 0 (no lsb) should always be greater in comparison
+              if (unsignedCompare(thislsb - 1, thatlsb - 1)) {
```

Comment:
Off-topic: I wonder if `unsignedCompare` (for `Int`) was ever benchmarked against a more straight-forward version that performs unsigned conversions to `Long`. It's also confusingly named (compare? for what?)

Reply:
I think (based in the HashSet changes, this can probably go from this implementation later

Reply:
removed reference to unsignedCompare from HashMap. There are still some from HashSet until https://github.com/scala/scala/pull/8589 merges